### PR TITLE
add TensorBase.is_complex

### LINKED ISSUE #339: Implement TensorBase.is_complex
### 🚀 Feature

Implement [`torch.is_complex`](https://pytorch.org/docs/stable/generated/torch.is_complex.html)

### Motivation

NeMo's [MegatronNeVaModel](https://github.com/NVIDIA/NeMo/blob/main/examples/multimodal/multimodal_llm/neva/neva_pretrain.py)

cc @apaz-cli @tfogal

### PR DESCRIPTION
**Pull request by khushi-411**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #339

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Sure thing!!!

### CODE REVIEW

File: thunder/tests/opinfos.py [hunk 1]
```diff
@@ -617,6 +617,7 @@ def _abs_torch(x: torch.Tensor | Number):
         ),
     ),
 )
+elementwise_unary_ops.append(abs_opinfo)
```
Comment (mruberry) [inline thunder/tests/opinfos.py:620]: Great catch! Thank you!

File: thunder/tests/opinfos.py [hunk 2]
```diff
@@ -1090,6 +1091,14 @@ def _abs_torch(x: torch.Tensor | Number):
 )
 elementwise_unary_ops.append(floor_opinfo)
 
+is_complex_opinfo = OpInfo(
```
Comment (mruberry) [inline thunder/tests/opinfos.py:1094]: `is_complex` is more tensor property like `is_cuda`, so let's put its OpInfo with the other tensor properties:

https://github.com/khushi-411/lightning-thunder/blob/05524ae8b828640587f6a67c72cd0c78671df649/thunder/tests/opinfos.py#L548

File: thunder/torch/__init__.py [hunk 3]
```diff
@@ -1102,6 +1102,11 @@ def ceil(a):
     return clang.ceil(a)
 
 
+@torchsymbol(torch.is_complex, is_method=True)
```
Comment (mruberry) [inline thunder/torch/__init__.py:1105]: Let's not make this a symbol and let's put it next to `is_cuda`:

https://github.com/khushi-411/lightning-thunder/blob/05524ae8b828640587f6a67c72cd0c78671df649/thunder/torch/__init__.py#L191

But then we need to two things (I think) to make it work properly:

- call `register_method()`
- update the map `_torch_to_thunder_function_map[torch.is_complex] = is_complex`

Then I would expect the function and method to work as expected

**Comment by mruberry** [comment]
Hey @khushi-411! I made some comments about how to add this as a non-symbol function and method. Let's see how that works! We haven't done a lot of that before, and other operators would benefit from a pattern for it, too

File: thunder/torch/__init__.py (hunk 3 above)
Comment (tfogal) [reply thunder/torch/__init__.py:1207]: Thanks for taking this on, @khushi-411! Mike had me do the same in #429, if you want more detail here.

Note that if it's both a method and a function of the module root (i.e. `torch.is_complex`) then AFAICT one needs to register both `torch.is_complex` and `torch.Tensor.is_complex`.

Comment (khushi-411) [reply thunder/torch/__init__.py:1262]: Hi, @tfogal! Thanks very much for the heads-up. I'll look into the PR and keep you posted.

I'm currently travelling for a week, but I will complete this PR soon. I apologize for the delay.

**Comment by riccardofelluga** [comment]
Picking this up since a month has passed without updates #644

**Comment by khushi-411** [comment]
> Picking this up since a month has passed without updates #644

Hi, @riccardofelluga! If you would like to take over the issue and work on it, please feel free to do so. I appreciate your understanding and assure you that I'll update the PRs quickly in the future. I apologize for the delay. Thank you!

**Pull request closed without merging**